src/Interpreters/Cache/Metadata.cpp

```diff
@@ -249,91 +233,202 @@ LockedKeyPtr CacheMetadata::lockKeyMetadata(
     return lockKeyMetadata(key, key_not_found_policy);
 }
 
-void CacheMetadata::iterate(IterateCacheMetadataFunc && func)
+void CacheMetadata::iterate(IterateFunc && func)
 {
     auto lock = lockMetadata();
-    for (const auto & [key, key_metadata] : *this)
+    for (auto & [key, key_metadata] : *this)
     {
-        LockedKeyPtr locked_key;
-        {
-            ProfileEventTimeIncrement<Microseconds> watch(ProfileEvents::FilesystemCacheLockKeyMicroseconds);
-            locked_key = std::make_unique<LockedKey>(key_metadata);
-        }
-
+        auto locked_key = key_metadata->lockNoStateCheck();
         const auto key_state = locked_key->getKeyState();
 
         if (key_state == KeyMetadata::KeyState::ACTIVE)
         {
             func(*locked_key);
             continue;
         }
-
-        if (key_state == KeyMetadata::KeyState::REMOVING)
+        else if (key_state == KeyMetadata::KeyState::REMOVING)
             continue;
 
         throw Exception(
             ErrorCodes::LOGICAL_ERROR, "Cannot lock key {}: key does not exist", key_metadata->key);
     }
 }
 
-void CacheMetadata::doCleanup()
+void CacheMetadata::removeAllKeys(bool if_releasable)
 {
     auto lock = lockMetadata();
+    for (auto it = begin(); it != end();)
+    {
+        auto locked_key = it->second->lockNoStateCheck();
+        if (locked_key->getKeyState() == KeyMetadata::KeyState::ACTIVE)
+        {
+            bool removed_all = locked_key->removeAllFileSegments(if_releasable);
+            if (removed_all)
+            {
+                it = removeEmptyKey(it, *locked_key, lock);
+                continue;
+            }
+        }
+        ++it;
+    }
+}
+
+void CacheMetadata::removeKey(const Key & key, bool if_exists, bool if_releasable)
+{
+    auto metadata_lock = lockMetadata();
 
-    FileCacheKey cleanup_key;
-    while (cleanup_queue->tryPop(cleanup_key))
+    auto it = find(key);
+    if (it == end())
     {
-        auto it = find(cleanup_key);
-        if (it == end())
-            continue;
+        if (if_exists)
+            return;
+        else
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "No such key: {}", key);
+    }
+
+    auto locked_key = it->second->lockNoStateCheck();
+    if (locked_key->getKeyState() != KeyMetadata::KeyState::ACTIVE)
+    {
+        if (if_exists)
+            return;
+        else
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "No such key: {}", key);
```

Comment:
It looks strange to throw "no such key" when the key exists and ACTIVE. Let's change the message?

Reply:
Isn't it better to say that it's a non-active state, then? So, it'd be easy to distinguish a previous logical error with the same message from this one